runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+---@class (private) vim.lsp.folding_range.BufState
+---
+---@field version? integer
+---
+--- Index in the form of row -> foldlevel
+---@field row_level table<integer, [integer, ">" | "<"?]?>
+---
+--- Index in the form of start_row -> collapsed_text
+---@field row_text table<integer, string?>
+---
+--- Index in the form of kind -> range
+---@field kind_ranges table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+
+---@type table<integer, vim.lsp.folding_range.BufState?>
+local bufstates = {}
+
+--- Update visible folds in the buffer.
+---@param bufnr integer
+local function foldupdate(bufnr)
+  for _, winid in ipairs(vim.fn.win_findbuf(bufnr)) do
+    local wininfo = vim.fn.getwininfo(winid)[1]
+    if wininfo and wininfo.tabnr == vim.fn.tabpagenr() then
+      if vim.wo[winid].foldmethod == 'expr' then
+        vim._foldupdate(winid, 0, api.nvim_buf_line_count(bufnr))
+      end
+    end
+  end
+end
+
+local augroup_update = api.nvim_create_augroup('vim_lsp_folding_range/update', {})
+
+--- Schedule `foldupdate()` after leaving insert mode.
+---@param bufnr integer
+local function schedule_foldupdate(bufnr)
+  if #(api.nvim_get_autocmds({
+    group = augroup_update,
+    buffer = bufnr,
+  })) == 0 then
+    api.nvim_create_autocmd('InsertLeave', {
+      group = augroup_update,
+      buffer = bufnr,
+      once = true,
+      callback = function()
+        foldupdate(bufnr)
+      end,
+    })
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/foldingRange`
+--- Store folding range then update folds for a specific buffer.
+---@private
+---@param result lsp.FoldingRange[]?
+---@param ctx lsp.HandlerContext
+local function handler(err, result, ctx)
+  if err then
+    log.error(err)
+    return
+  end
+
+  if not result then
+    return
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  ---@type table<integer, [integer, ">" | "<"?]?>
+  local row_level = {}
+  ---@type table<integer, string>
+  local row_text = {}
+  ---@type table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+  local kind_ranges = {}
+
+  for _, range in ipairs(result) do
+    local kind = range.kind
+    if kind then
+      ---@cast kind lsp.FoldingRangeKind
```

Comment:
Is LuaLS not able to infer this?

Reply:
I remember it wasn't before, but I am about to change the data structure of `kind_ranges` (for multi-client support, and to ensure it collapses from the smallest place, this issue will not exist).